File: oltrack-ts/src/adapters/dhl.ts
import type { Location, Shipment, TrackingEvent, UlscCode } from "../types.js";
import { event, shipment } from "../normalize.js";

/**
 * Subset of DHL event:ric → UlscCode mapping (from mappings/intl/dhl.csv).
 *
 * v0.5 MVP: inline ~25 most common combinations. v1.0 will auto-generate
 * the full 282-row table at build time.
 */
const DHL_RAW_TO_ULSC: Record<string, UlscCode> = {
  // status:* (top-level statusCode)
  "status:pre-transit": "order_created",
  "status:transit": "in_transit",
  "status:delivered": "delivered",
  "status:failure": "exception",
  "status:unknown": "exception",
  // event:ric — common ones from parcel-de
  "PCKDU:PUBCR": "picked_up",
  "PCKDU:PFLOC": "picked_up",
  "DLVRD:ACCPT": "delivered",
  "DLVRD:ACCDM": "damaged",
  "DLVRD:CUSTM": "customs_declared",
  "DLVRD:NGHBR": "signed_by_third_party",
  "DLVRD:SERPT": "delivery_to_locker",
  "DLVRD:HNDTC": "transferred_to_carrier",
  "DLVRF:NTORD": "refused",
  "DLVRF:DMGED": "damaged",
  "MVARR:NRQRD": "arrived_at_hub",
  "MVDPT:NRQRD": "departed_from_hub",
  "ULFMV:OARRV": "arrived_at_hub",
  "INFCL:NRQRD": "in_transit",
  "INFCL:OARRV": "arrived_at_hub",
  "SRTED:NRQRD": "in_transit",
  "ICPRC:NRQRD": "customs_declared",
  "ICPRC:OARRV": "customs_declared",
  "RSDCP:OARRV": "customs_released",
  "HLDCU:NORSN": "customs_held",
  "CNRFC:HDFCT": "awaiting_pickup",
  "RETRN:RFUSD": "return_initiated",
};

interface DhlAddress {
  addressLocality?: string;
  countryCode?: string;
  postalCode?: string;
}

interface DhlEvent {
  timestamp: string;
  statusCode?: string;
  status?: string;  // 在 events[] 里 status 通常是 event:ric form
  description?: string;
  remark?: string;
  location?: { address?: DhlAddress };
  pieceId?: string;
}

interface DhlShipment {
  trackingNumber: string;
  service?: string;
  origin?: { address?: DhlAddress };
  destination?: { address?: DhlAddress };
  estimatedTimeOfDelivery?: string;
  events?: DhlEvent[];
}

interface DhlResponse {
  shipments?: DhlShipment[];
}


function parseLocation(loc?: { address?: DhlAddress }): Location | undefined {
  if (!loc) return undefined;
  const addr = loc.address ?? {};
  const out: Location = {};
  if (addr.addressLocality) {
    out.name = addr.addressLocality;
    // addressLocality is "CITY - STATE - COUNTRY"; first segment 当 city
    const parts = addr.addressLocality.split("-").map((p) => p.trim());
    if (parts[0]) out.city = parts[0];
  }
  if (addr.countryCode) out.countryCode = addr.countryCode;
  if (addr.postalCode) out.postalCode = addr.postalCode;
  return Object.keys(out).length > 0 ? out : undefined;
}


export function normalizeEvent(ev: DhlEvent): TrackingEvent {
  // Prefer event-level event:ric in `status`, fall back to statusCode 加 "status:" 前缀
  let carrierEventCode: string;
  if (ev.status && ev.status.includes(":")) {
    carrierEventCode = ev.status;
  } else if (ev.statusCode) {
    carrierEventCode = `status:${ev.statusCode}`;
  } else {
    throw new Error("DHL event missing both status and statusCode");
  }

  const oltsCode = DHL_RAW_TO_ULSC[carrierEventCode];
  if (!oltsCode) {
    throw new Error(
      `unknown DHL raw code ${carrierEventCode}; extend DHL_RAW_TO_ULSC or fall back to mappings/intl/dhl.csv lookup`
    );
  }

  return event({
    carrierCode: "dhl",
    carrierEventCode,
    eventTime: ev.timestamp,
    oltsCode,
    description: ev.description ?? ev.remark,
    location: parseLocation(ev.location),
    pieceId: ev.pieceId,
  });
}


export function normalizeResponse(raw: DhlResponse): TrackingEvent[] {
  const shipments = raw.shipments ?? [];
  if (shipments.length === 0) return [];
  const events = shipments[0]!.events ?? [];
  return events.map(normalizeEvent);
}


export function normalizeToShipment(raw: DhlResponse): Shipment {
  const shipments = raw.shipments ?? [];
  if (shipments.length === 0) throw new Error("DHL response has no shipments");
  const sh = shipments[0]!;

  const events = (sh.events ?? []).map(normalizeEvent);
  const origin = sh.origin?.address?.countryCode
    ? { countryCode: sh.origin.address.countryCode }
    : undefined;
  const destination = sh.destination?.address?.countryCode
    ? { countryCode: sh.destination.address.countryCode }
    : undefined;

  return shipment({
    trackingNumber: sh.trackingNumber,
    carrierCode: "dhl",
    events,
    service: sh.service,
    origin,
    destination,
    estimatedDelivery: sh.estimatedTimeOfDelivery,
  });
}

File: oltrack-py/src/oltrack/adapters/dhl.py
"""Adapter for DHL Tracking Unified API (api-eu.dhl.com).

Reference response shape (per mappings/intl/dhl.csv source):

    {
      "shipments": [{
        "shipmentId": "0034988765432",
        "trackingNumber": "0034988765432",
        "service": "express",
        "status": {
          "statusCode": "delivered",
          "status": "Delivered",
          "timestamp": "2024-06-04T03:17:00+02:00"
        },
        "estimatedTimeOfDelivery": "2024-06-05",
        "origin": {"address": {"countryCode": "CN"}},
        "destination": {"address": {"countryCode": "DE"}},
        "events": [
          {
            "timestamp": "2024-06-04T03:17:00+02:00",
            "statusCode": "transit",
            "status": "DLVRD:CUSTM",  // carrier_event_code form
            "description": "Handed over to customs",
            "location": {"address": {"addressLocality": "FRANKFURT - DE", "countryCode": "DE"}}
          },
          ...
        ]
      }]
    }

DHL exposes both a top-level StatusCode (5 values) and event-level event:ric
codes. We use the event-level when available, falling back to the statusCode
prefix.
"""

from __future__ import annotations

from typing import Any

from oltrack import normalize


def _parse_location(loc: dict[str, Any]) -> dict[str, Any]:
    """Parse DHL location subobject into OLTS Location."""
    if not loc:
        return {}
    address = loc.get("address", {}) if isinstance(loc, dict) else {}
    out: dict[str, Any] = {}
    if addr_loc := address.get("addressLocality"):
        out["name"] = addr_loc
        # addressLocality is often "CITY - STATE - COUNTRY"; take first as city
        parts = [p.strip() for p in addr_loc.split("-")]
        if parts:
            out["city"] = parts[0]
    if cc := address.get("countryCode"):
        out["countryCode"] = cc
    if pc := address.get("postalCode"):
        out["postalCode"] = pc
    return out


def normalize_event(event_obj: dict[str, Any]) -> dict[str, Any]:
    """Convert one events[] item to an OLTS TrackingEvent."""
    # Prefer event-level event:ric code in `status`; fall back to statusCode with "status:" prefix
    raw_status = event_obj.get("status")
    if raw_status and ":" in raw_status:
        carrier_event_code = raw_status
    else:
        carrier_event_code = f"status:{event_obj.get('statusCode', 'unknown')}"

    location = _parse_location(event_obj.get("location") or {})
    return normalize.event(
        carrier="dhl",
        carrier_event_code=carrier_event_code,
        event_time=event_obj["timestamp"],
        description=event_obj.get("description") or event_obj.get("remark"),
        location=location or None,
        piece_id=event_obj.get("pieceId"),
        metadata={k: v for k, v in event_obj.items()
                  if k in ("statusDetailed", "nextSteps") and v},
    )


def normalize_response(raw: dict[str, Any]) -> list[dict[str, Any]]:
    """Parse a DHL Tracking Unified API response into a flat list of events."""
    shipments = raw.get("shipments", [])
    if not shipments:
        return []
    return [normalize_event(e) for e in shipments[0].get("events", [])]


def normalize_to_shipment(raw: dict[str, Any]) -> dict[str, Any]:
    """Parse a full DHL response into an OLTS Shipment dict."""
    shipments = raw.get("shipments", [])
    if not shipments:
        raise ValueError("DHL response has no shipments")
    sh = shipments[0]
    events = [normalize_event(e) for e in sh.get("events", [])]
    origin_addr = (sh.get("origin") or {}).get("address", {})
    dest_addr = (sh.get("destination") or {}).get("address", {})
    origin = {"countryCode": origin_addr["countryCode"]} if origin_addr.get("countryCode") else None
    destination = {"countryCode": dest_addr["countryCode"]} if dest_addr.get("countryCode") else None

    return normalize.shipment(
        tracking_number=sh["trackingNumber"],
        events=events,
        carrier="dhl",
        service=sh.get("service"),
        origin=origin,
        destination=destination,
        estimatedDelivery=sh.get("estimatedTimeOfDelivery"),
    )
